SteamKit2/SteamKit2/Util/DebugNetworkListener.cs

```diff
@@ -0,0 +1,115 @@
+﻿/*
+ * This file is subject to the terms and conditions defined in
+ * file 'license.txt', which is part of this source code package.
+ */
+
+
+
+using System;
+using System.IO;
+using System.Threading;
+
+namespace SteamKit2
+{
+    /// <summary>
+    /// This is a debug utility, do not use it to implement your business logic.
+    /// 
+    /// This interface is used for logging network messages sent to and received from the Steam server that the client is connected to.
+    /// </summary>
+    public interface IDebugNetworkListener
+    {
+        /// <summary>
+        /// Called when a packet is received from the Steam server.
+        /// </summary>
+        /// <param name="msgType">Network message type of this packet message.</param>
+        /// <param name="data">Raw packet data that was received.</param>
+        void OnIncomingNetMessage( EMsg msgType, byte[] data );
+
+        /// <summary>
+        /// Called when a packet is about to be sent to the Steam server.
+        /// </summary>
+        /// <param name="msgType">Network message type of this packet message.</param>
+        /// <param name="data">Raw packet data that will be sent.</param>
+        void OnOutgoingNetMessage( EMsg msgType, byte[] data );
+    }
+
+    /// <summary>
+    /// Dump any network messages sent to and received from the Steam server that the client is connected to.
+    /// These messages are dumped to file, and can be analyzed further with NetHookAnalyzer, a hex editor, or your own purpose-built tools.
+    /// </summary>
+    public class NetHookNetworkListener : IDebugNetworkListener
+    {
+        private long MessageNumber;
+        private string LogDirectory;
+
+        /// <summary>
+        /// Will create a folder in path "%assembly%/nethook/%currenttime%/"
+        /// </summary>
+        public NetHookNetworkListener()
+        {
+            var uri = new Uri( System.Reflection.Assembly.GetExecutingAssembly().CodeBase );
+
+            LogDirectory = Path.Combine(
+                Path.GetDirectoryName( uri.LocalPath ),
+                "nethook",
+                DateUtils.DateTimeToUnixTime( DateTime.Now ).ToString()

```

Comment:
Seems a bit racy - two concurrently created `SteamClient`s with separate listeners could both try claim the same directory.